core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/XmlWriter.java

```diff
@@ -193,6 +193,8 @@ private String escapeXmlEntities(String s) {
             s = s.replace("&apos;", "'");
             s = s.replace("&lt;", "<");
             s = s.replace("&gt;", ">");
+            s = s.replace("&#x0D;", "\r");
+            s = s.replace("&#x0A;", "\n");
```

Comment:
This feels super inefficient, but I guess it's just 1.x. We should do better in 2.x to avoid all of the string copying.

Reply:
+1 The micro optimization articles I shared a while ago come handy now! :wink:

Reply:
Thanks Matt and Zoe , will add this optimization in another PR 
https://medium.com/javarevisited/micro-optimizations-in-java-string-replaceall-c6d0edf2ef6